mypy/semanal.py

```diff
@@ -581,7 +581,11 @@ def refresh_partial(
     def refresh_top_level(self, file_node: MypyFile) -> None:
         """Reanalyze a stale module top-level in fine-grained incremental mode."""
         self.recurse_into_functions = False
-        self.add_implicit_module_attrs(file_node)
+        # We do it in the very last order, because of
+        # `builtins.dict <-> typing.Mapping <-> abc.ABCMeta`
+        # cyclic imports.
```

Comment:
The problem is not the cyclic import itself, but that the method assumes that e.g. `dict` (for `__annotations__`) would be never marked incomplete (i.e. it may be not defined yet, but as soon as it is, it will be `TypeInfo`, rather than `PlaceholderNode`). This makes me think maybe a cleaner solution (if it works), would be just to replace `assert` with an `if` (and simply continue if it is a placeholder, like we do if the symbol is `None`). Can you try this?

Btw curiously this seem to only affect tests for now, it looks in real typeshed currently `dict` is not an instance of `ABCMeta`.

Reply:
> Can you try this?

Done, looks like it works! Let's wait to see what tests will show us :)